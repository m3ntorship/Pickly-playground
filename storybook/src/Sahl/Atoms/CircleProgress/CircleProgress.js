import React, { useEffect, useState, useRef } from "react";
import PropTypes from "prop-types";

const CircleProgress = (props) => {
  const [offset, setOffset] = useState(0);
  const circleRef = useRef(null);
  const {
    size = 72,
    progress,
    strokeWidth = 7,
    backgroundColor = "#fff",
    circleColor = "#7048E8",
  } = props;

  const center = size / 2;
  const radius = size / 2 - strokeWidth / 2;
  const circumference = 2 * Math.PI * radius;

  useEffect(() => {
    const progressOffset = ((100 - progress) / 100) * circumference;
    setOffset(progressOffset);

    // circleRef.current.style = "transition: stroke-dashoffset 850ms ease-in-out";.transition
    if (circleRef.current) {
      circleRef.current.style.transition =
        "stroke-dashoffset 850ms ease-in-out";
      circleRef.current.style.fill = `${backgroundColor}`;
    }
  }, [setOffset, progress, circumference, offset, backgroundColor]);

  return (
    <div className="relative inline-block">
      <svg className={`inline-block `} width={size} height={size}>
        {/* Background Circle */}
        <circle
          stroke="none"
          fill="none"
          cx={center}
          cy={center}
          r={radius}
          strokeWidth={4}
        />
        {/* Outline Circle */}
        <circle
          ref={circleRef}
          stroke={circleColor}
          fill="none"
          cx={center}
          cy={center}
          r={radius}
          strokeWidth={strokeWidth}
          strokeDasharray={circumference}
          strokeDashoffset={offset}
        />
      </svg>
      <span className="inline-block w-min absolute left-1/2 top-2/3 transform -translate-x-2/4 -translate-y-2/3 text-dark text-sm">
        {progress}%
      </span>
    </div>
  );
};
// left-0 right-0 mx-auto
CircleProgress.propTypes = {
  size: PropTypes.number,
  progress: PropTypes.number.isRequired,
  strokeWidth: PropTypes.number,
  backgroundColor: PropTypes.string,
  circleColor: PropTypes.string,
};

export default CircleProgress;
