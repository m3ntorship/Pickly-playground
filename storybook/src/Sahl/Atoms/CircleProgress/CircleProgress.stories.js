import React from "react";

import CircleProgress from "./CircleProgress";

const defaultExport = {
  title: "Sahl/Atoms/CircleProgress",
  component: CircleProgress,
  argTypes: {
    // Default without options
    // progress: { control: "range" },
    progress: {
      control: {
        type: "range",
        min: 0,
        max: 100,
        step: 5,
      },
    },
    circleColor: { control: "color" },
    backgroundColor: { control: "color" },
  },
};

export default defaultExport;

const Template = (args) => <CircleProgress {...args} />;

export const Default = Template.bind({});
Default.args = {
  progress: 40,
  strokeWidth: 7,
  size: 72,
};
