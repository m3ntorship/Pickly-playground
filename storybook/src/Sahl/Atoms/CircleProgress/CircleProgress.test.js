import React from "react";
import renderer from "react-test-renderer";
import CircleProgress from "./CircleProgress";

describe("CircleProgress Component", () => {
  it("should give error when no props are passed", () => {
    const tree = renderer.create(<CircleProgress />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should render correctly when only progress prop is passed", () => {
    const tree = renderer.create(<CircleProgress progress={50} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should change background color when background is passed as color string", () => {
    const tree = renderer
      .create(<CircleProgress progress={50} backgroundColor="#819" />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should change circle color when circle color is passed as color string", () => {
    const tree = renderer
      .create(<CircleProgress progress={50} circleColor="#812" />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should change circle radius when passing size", () => {
    const tree = renderer
      .create(<CircleProgress progress={50} size={32} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
