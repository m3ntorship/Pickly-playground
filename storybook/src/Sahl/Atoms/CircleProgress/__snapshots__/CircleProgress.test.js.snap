// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CircleProgress Component should change background color when background is passed as color string 1`] = `
<div
  className="relative inline-block"
>
  <svg
    className="inline-block "
    height={72}
    width={72}
  >
    <circle
      cx={36}
      cy={36}
      fill="none"
      r={32.5}
      stroke="none"
      strokeWidth={4}
    />
    <circle
      cx={36}
      cy={36}
      fill="none"
      r={32.5}
      stroke="#7048E8"
      strokeDasharray={204.20352248333654}
      strokeDashoffset={0}
      strokeWidth={7}
    />
  </svg>
  <span
    className="inline-block w-min absolute left-1/2 top-2/3 transform -translate-x-2/4 -translate-y-2/3 text-dark text-sm"
  >
    50
    %
  </span>
</div>
`;

exports[`CircleProgress Component should change circle color when circle color is passed as color string 1`] = `
<div
  className="relative inline-block"
>
  <svg
    className="inline-block "
    height={72}
    width={72}
  >
    <circle
      cx={36}
      cy={36}
      fill="none"
      r={32.5}
      stroke="none"
      strokeWidth={4}
    />
    <circle
      cx={36}
      cy={36}
      fill="none"
      r={32.5}
      stroke="#812"
      strokeDasharray={204.20352248333654}
      strokeDashoffset={102.10176124166827}
      strokeWidth={7}
    />
  </svg>
  <span
    className="inline-block w-min absolute left-1/2 top-2/3 transform -translate-x-2/4 -translate-y-2/3 text-dark text-sm"
  >
    50
    %
  </span>
</div>
`;

exports[`CircleProgress Component should change circle radius when passing size 1`] = `
<div
  className="relative inline-block"
>
  <svg
    className="inline-block "
    height={32}
    width={32}
  >
    <circle
      cx={16}
      cy={16}
      fill="none"
      r={12.5}
      stroke="none"
      strokeWidth={4}
    />
    <circle
      cx={16}
      cy={16}
      fill="none"
      r={12.5}
      stroke="#7048E8"
      strokeDasharray={78.53981633974483}
      strokeDashoffset={0}
      strokeWidth={7}
    />
  </svg>
  <span
    className="inline-block w-min absolute left-1/2 top-2/3 transform -translate-x-2/4 -translate-y-2/3 text-dark text-sm"
  >
    50
    %
  </span>
</div>
`;

exports[`CircleProgress Component should give error when no props are passed 1`] = `
<div
  className="relative inline-block"
>
  <svg
    className="inline-block "
    height={72}
    width={72}
  >
    <circle
      cx={36}
      cy={36}
      fill="none"
      r={32.5}
      stroke="none"
      strokeWidth={4}
    />
    <circle
      cx={36}
      cy={36}
      fill="none"
      r={32.5}
      stroke="#7048E8"
      strokeDasharray={204.20352248333654}
      strokeDashoffset={0}
      strokeWidth={7}
    />
  </svg>
  <span
    className="inline-block w-min absolute left-1/2 top-2/3 transform -translate-x-2/4 -translate-y-2/3 text-dark text-sm"
  >
    %
  </span>
</div>
`;

exports[`CircleProgress Component should render correctly when only progress prop is passed 1`] = `
<div
  className="relative inline-block"
>
  <svg
    className="inline-block "
    height={72}
    width={72}
  >
    <circle
      cx={36}
      cy={36}
      fill="none"
      r={32.5}
      stroke="none"
      strokeWidth={4}
    />
    <circle
      cx={36}
      cy={36}
      fill="none"
      r={32.5}
      stroke="#7048E8"
      strokeDasharray={204.20352248333654}
      strokeDashoffset={102.10176124166827}
      strokeWidth={7}
    />
  </svg>
  <span
    className="inline-block w-min absolute left-1/2 top-2/3 transform -translate-x-2/4 -translate-y-2/3 text-dark text-sm"
  >
    50
    %
  </span>
</div>
`;
